src/compiler/parser.ts

```diff
@@ -5698,9 +5700,10 @@ namespace ts {
             const hasJSDoc = hasPrecedingJSDocComment();
             parseExpected(SyntaxKind.WhileKeyword);
             const openParenPosition = scanner.getTokenPos();
-            parseExpected(SyntaxKind.OpenParenToken);
+            const openParenExists = parseExpected(SyntaxKind.OpenParenToken);
             const expression = allowInAnd(parseExpression);
-            parseExpectedMatchingBrackets(SyntaxKind.OpenParenToken, SyntaxKind.CloseParenToken, openParenPosition);
+            openParenExists ? parseExpectedMatchingBrackets(SyntaxKind.OpenParenToken, SyntaxKind.CloseParenToken, openParenPosition)

```

Comment:
I think it's important for this subtle pitfall to be fixed in one place, `parseExpectedMatchingBrackets`, and for it to require its callers to provide it enough information to do so. I'm not sure of the most elegant way though. Maybe an extra parameter, or maybe changing parseExpected to return `number | undefined` instead of `boolean`. But the second option is kind of arbitrary and useless to all its many other callers.